test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/EnumTest.java

```diff
@@ -0,0 +1,96 @@
+ */
+
+package software.amazon.awssdk.protocol.tests;
+
+import static java.util.stream.Collectors.toMap;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Stream;
+import org.junit.Test;
+import software.amazon.awssdk.services.protocolrestjson.model.AllTypesResponse;
+import software.amazon.awssdk.services.protocolrestjson.model.EnumType;
+
+/**
+ * Verifies that the behavior when retrieving enums from a model behaves as expected.
+ */
+public class EnumTest {
+    @Test
+    public void knownEnumFieldsBehaveCorrectly() {
+        assertThat(simulateUnmarshallingEnum("EnumValue1")).isEqualTo(EnumType.ENUM_VALUE1);
+        assertThat(simulateUnmarshallingEnum((String) null)).isEqualTo(null);
+        assertThat(simulateUnmarshallingEnum(EnumType.ENUM_VALUE1)).isEqualTo(EnumType.ENUM_VALUE1);
+    }
+
+    @Test
+    public void unknownEnumFieldsBehaveCorrectly() {
+        assertThat(simulateUnmarshallingEnum("Foo")).isEqualTo(EnumType.UNKNOWN);
+    }
+
+    @Test
+    public void knownEnumListFieldsBehaveCorrectly() {
+        assertThat(simulateUnmarshallingEnumList("EnumValue1", "EnumValue2"))
+                .containsExactly(EnumType.ENUM_VALUE1, EnumType.ENUM_VALUE2);
+        assertThat(simulateUnmarshallingEnumList(new String[] { null })).containsExactly(new EnumType[] { null });
+        assertThat(simulateUnmarshallingEnumList()).isEmpty();
+    }
+
+    @Test
+    public void unknownEnumListFieldsBehaveCorrectly() {
+        assertThat(simulateUnmarshallingEnumList("Foo", "EnumValue2")).containsExactly(EnumType.UNKNOWN, EnumType.ENUM_VALUE2);
+    }
+
+    @Test
+    public void knownEnumMapFieldsBehaveCorrectly() {
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("EnumValue1", "EnumValue2")))
+                .containsExactly(new SimpleImmutableEntry<>(EnumType.ENUM_VALUE1, EnumType.ENUM_VALUE2));
+    }
+
+    @Test
+    public void unknownEnumMapFieldsBehaveCorrectly() {
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("Foo", "EnumValue2")))
+                .isEmpty();
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("Foo", "Foo")))
+                .isEmpty();
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("Foo", "")))
+                .isEmpty();
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("EnumValue1", "Foo")))
+                .containsExactly(new SimpleImmutableEntry<>(EnumType.ENUM_VALUE1, EnumType.UNKNOWN));
+        assertThat(simulateUnmarshallingEnumMap(new SimpleImmutableEntry<>("EnumValue1", "")))
+                .containsExactly(new SimpleImmutableEntry<>(EnumType.ENUM_VALUE1, EnumType.UNKNOWN));
+    }
+
+    private EnumType simulateUnmarshallingEnum(String value) {
```

Comment:
Seems kind of a weird name. You're really just testing the builder behavior here.

Reply:
I'm doing what the unmarshaller does. I'll add a comment.